### ISSUE TITLE
MSGPACK_ADD_ENUM is not working  with c++11

### ISSUE THREAD

**Issue by davidchappelle** [msgpack/msgpack-c#248]
I originally submitted the defect for:

    https://github.com/msgpack/msgpack-c/issues/204

This was patched and resolved on Jan 31 via:

    https://github.com/msgpack/msgpack-c/commit/c4fb47c00dccaedb238787e662c5262b12aa70c7

When I build my code preceding this commit everything builds as expected. When I build my code with this commit my build is broken and I get the following compilation errors:

```
dchappelle@orion:build$ make
Scanning dependencies of target bonefish
[  4%] Building CXX object src/CMakeFiles/bonefish.dir/bonefish/broker/wamp_broker.cpp.o
[  8%] Building CXX object src/CMakeFiles/bonefish.dir/bonefish/common/wamp_message_processor.cpp.o
[ 12%] Building CXX object src/CMakeFiles/bonefish.dir/bonefish/dealer/wamp_dealer.cpp.o
[ 16%] Building CXX object src/CMakeFiles/bonefish.dir/bonefish/messages/wamp_message_defaults.cpp.o
[ 20%] Building CXX object src/CMakeFiles/bonefish.dir/bonefish/messages/wamp_message_factory.cpp.o
[ 25%] Building CXX object src/CMakeFiles/bonefish.dir/bonefish/messages/wamp_message_type.cpp.o
[ 29%] Building CXX object src/CMakeFiles/bonefish.dir/bonefish/messages/wamp_welcome_details.cpp.o
[ 33%] Building CXX object src/CMakeFiles/bonefish.dir/bonefish/router/wamp_router.cpp.o
In file included from /home/dchappelle/environment/bonefish/third-party/msgpack-c/include/msgpack.hpp:18:0,
                 from /home/dchappelle/environment/bonefish/src/bonefish/messages/wamp_welcome_details.hpp:31,
                 from /home/dchappelle/environment/bonefish/src/bonefish/router/wamp_router.hpp:5,
                 from /home/dchappelle/environment/bonefish/src/bonefish/router/wamp_router.cpp:1:
/home/dchappelle/environment/bonefish/third-party/msgpack-c/include/msgpack/object.hpp: In instantiation of ‘void msgpack::v1::operator<<(msgpack::v1::object::with_zone&, const T&) [with T = bonefish::wamp_message_type]’:
/home/dchappelle/environment/bonefish/third-party/msgpack-c/include/msgpack/object.hpp:514:30:   required from ‘msgpack::v1::object::object(const T&, msgpack::v1::zone*) [with T = bonefish::wamp_message_type]’
/home/dchappelle/environment/bonefish/src/bonefish/messages/wamp_abort_message.hpp:50:16:   required from here
/home/dchappelle/environment/bonefish/third-party/msgpack-c/include/msgpack/object.hpp:311:5: error: request for member ‘msgpack_object’ in ‘v’, which is of non-class type ‘const bonefish::wamp_message_type’
     v.msgpack_object(static_cast<object*>(&o), o.zone);
     ^
/home/dchappelle/environment/bonefish/third-party/msgpack-c/include/msgpack/object.hpp: In instantiation of ‘const msgpack::v1::object& msgpack::v1::operator>>(const msgpack::v1::object&, T&) [with T = bonefish::wamp_message_type]’:
/home/dchappelle/environment/bonefish/third-party/msgpack-c/include/msgpack/object.hpp:429:33:   required from ‘void msgpack::v1::object::convert(T&) const [with T = bonefish::wamp_message_type]’
/home/dchappelle/environment/bonefish/third-party/msgpack-c/include/msgpack/object.hpp:442:5:   required from ‘T msgpack::v1::object::as() const [with T = bonefish::wamp_message_type]’
/home/dchappelle/environment/bonefish/src/bonefish/messages/wamp_abort_message.hpp:60:41:   required from here
/home/dchappelle/environment/bonefish/third-party/msgpack-c/include/msgpack/object.hpp:127:5: error: request for member ‘msgpack_unpack’ in ‘v’, which is of non-class type ‘bonefish::wamp_message_type’
     v.msgpack_unpack(o.convert());
     ^
make[2]: *** [src/CMakeFiles/bonefish.dir/bonefish/router/wamp_router.cpp.o] Error 1
make[1]: *** [src/CMakeFiles/bonefish.dir/all] Error 2
make: *** [all] Error 2
```

I noticed in the patch that the 'template<>' lines were removed from the MSGPACK_ADD_ENUM macro. So I added them back in and now my build succeeds as it did before. Seems that these functions are defined as template functions in object.hpp. Another oddity here is that the following function from the macro does not have a template definition in object.cpp:

```
inline void operator<< (msgpack::object& o, const enum& v)
```

Adding a 'template<>' to this function definintion results in an error. So the questions are as follows:

1. Why were the template<> lines removed from the macro to begin with?
2. Why doesn't __inline void operator<< (msgpack::object& o, const enum& v)__ have a corresponding templated declaration/definition in object.hpp?

I also wrote a simple main.cpp program with an enum and it seems to work. So why is this failing for me in my build? Why does adding in the template<> lines get the build working again? There is some strange voodoo happening here.

**Comment by redboltz** [comment]
Hi @davidchappelle, thank you for reporting the issue. As you noticed, the issue is relate to #249.

I answer two questions.

1. Why were the template<> lines removed from the macro to begin with?

When the header files includes correct order, function overloading works well. And in this case, I believe that the function overload is a better solution. See http://www.gotw.ca/gotw/049.htm

2. Why doesn't inline void operator<< (msgpack::object& o, const enum& v) have a corresponding templated declaration/definition in object.hpp?

It is caused by header files including order. I will write a comment in #249

**Comment by redboltz** [comment]
Fixed by #262
